---
title: "Process Genotypes For Ancestry Analysis"
author: "JKG"
date: "9/18/2019"
output: pdf_document
  chunk_output_type: console
---

```{r libs, include=FALSE}
knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)

```

## Pull 1000Genomes down and process
* Pull out reference populations (CEU, YRI, CHB, JPT)
```{bash pull_1000Genomes, echo=FALSE}
source ~/.bashrc
PATH=$PATH:/home/jgockley/TWAS/bin/

CORES=`expr $(nproc) - 2`

#ID Sample list of needed Ref Populations
wget ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/phase1/analysis_results/integrated_call_sets/integrated_call_samples.20101123.ALL.panel

ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/phase1/analysis_results/integrated_call_sets/

grep CEU integrated_call_samples.20101123.ALL.panel | cut -f1 > CEU.samples.list
grep YRI integrated_call_samples.20101123.ALL.panel | cut -f1 > YRI.samples.list
grep CHB integrated_call_samples.20101123.ALL.panel | cut -f1 > CHB.samples.list
grep JPT integrated_call_samples.20101123.ALL.panel | cut -f1 > JPT.samples.list

#Make Plink IDV List of all 1000G Ref Samps to pull
cat CEU.samples.list YRI.samples.list CHB.samples.list JPT.samples.list > 1000G_IDV.list
paste 1000G_IDV.list 1000G_IDV.list > foo.txt
mv foo.txt 1000G_IDV.list

##Pull The LDREF Panel SNPS
#Pull LDREF Data
wget https://data.broadinstitute.org/alkesgroup/FUSION/LDREF.tar.bz2
tar xjvf LDREF.tar.bz2

#Merge  LDREF Data
for i in {2..22} 
do 
    echo LDREF/1000G.EUR.$i>> MergedChrs.txt
done 
plink --bfile LDREF/1000G.EUR.1 --threads $CORES --merge-list MergedChrs.txt --make-bed --allow-no-sex --out TotalLDRef

#Write SNP List for LDREF Data
plink --threads $CORES --bfile TotalLDRef --write-snplist --out LDREF_SNPs

#Reformat SNP List for plink filter.txt
 awk '{ print $1"\t"$4"\t"$4"\t"$2 }' TotalLDRef.bim > LDREF_SNPs.txt

rm -r LDREF/

#Pull 1000G Data, convert to plink, Filt for INDV and SNPs
for i in {1..22}
do
	wget ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/phase1/analysis_results/integrated_call_sets/ALL.chr$i\.integrated_phase1_v3.20101123.snps_indels_svs.genotypes.vcf.gz

    plink --vcf ALL.chr$i\.integrated_phase1_v3.20101123.snps_indels_svs.genotypes.vcf.gz --snps-only --extract LDREF_SNPs.txt --keep 1000G_IDV.list --threads $CORES --make-bed --out 1000Genomes_chr$i
  rm ALL.chr$i\.integrated_phase1_v3.20101123.snps_indels_svs.genotypes.vcf.gz
done

#Merge All Plink Chr Files 
for i in {2..22}
do
  echo 1000Genomes_chr$i >> Merge.list
done

plink --bfile 1000Genomes_chr1 --merge-list Merge.list --make-bed --out 1000Genomes_Merged

rm 1000Genomes_chr*
rm *.list
rm *.panel
rm LDREF.tar.bz2
rm MergedChrs.txt
```

## Process MSBB WGS Data
* Pull VCFs from Synapse
* Convert to plink
* Merge across chromosomes
* Filter for 1000G SNPs
```{bash ProcessMSBB, echo=FALSE}
CORES=`expr $(nproc) - 2`
PATH=$PATH:/home/jgockley/TWAS/bin/
##Pull VCF Files from Synapse
mkdir VCF
synapse get -r syn11707204 --downloadLocation VCF/

##Convert VCFs to plink
mkdir GWAS_Plink

rm VCF/NIA_JG_1898_samples_GRM_WGS_b37_JointAnalysis01_2017-12-08_X*
rm VCF/NIA_JG_1898_samples_GRM_WGS_b37_JointAnalysis01_2017-12-08_Y*
rm VCF/NIA_JG_1898_samples_GRM_WGS_b37_JointAnalysis01_2017-12-08_others*
rm -r VCF/WGS_JointCalling/
rm VCF/SYNAPSE_METADATA_MANIFEST.tsv

#Convert to PLINK
for i in {22..1}
do
  plink --vcf VCF/NIA_JG_1898_samples_GRM_WGS_b37_JointAnalysis01_2017-12-08_$i\.recalibrated_variants.MSSM.vcf.gz --threads $CORES --make-bed --allow-no-sex --snps-only --out GWAS_Plink/MSSB_Chr_$i
  rm VCF/NIA_JG_1898_samples_GRM_WGS_b37_JointAnalysis01_2017-12-08_$i\.recalibrated_variants.MSSM.vcf.*
done
rm -r VCF/

#Replace the "." IDs with a uniq ID
for i in {22..1}
do
  awk '{ print $1"\t"$1"_"$4"\t"$3"\t"$4"\t"$5"\t"$6 }' GWAS_Plink/MSSB_Chr_$i\.bim > foo.txt
  mv foo.txt GWAS_Plink/MSSB_Chr_$i\.bim
done

#Make an LD_Ref SNPList with matching name Schema
awk '{print $1"\t"$2"\t"$3"\t"$1"_"$2 }' LDREF_SNPs.txt > MSBB_LD_SNPS.txt

##Merge Chromsomes in Plink Format 
for i in {2..22} 
do 
    echo GWAS_Plink/MSSB_Chr_$i >> MergedGWAS.txt
done 

CORES=`expr $(nproc) - 2`
PATH=$PATH:/home/jgockley/TWAS/bin/
plink --threads $CORES --bfile GWAS_Plink/MSSB_Chr_1 --merge-list MergedGWAS.txt --extract MSBB_LD_SNPS.txt --make-bed --allow-no-sex --out GWASTotal
#rm -r WGS_Plink/

#Rename Snps with rsIDs & Centimorgan Distance
#python BimAnotator.py TotalLDRef.bim LDREF_FiltSNPs.bim
```

## Process MAYO Genotyping Data
```{bash MAYO}
CORES=`expr $(nproc) - 2`
#Download
mkdir Mayo
synapse get -r syn8650955 --downloadLocation Mayo/

plink --bfile Mayo/MayoRNAseq_RNAseq_Genome-Wide_Genotypes_HRCimputed --threads $CORES --extract LDREF_SNPs.txt --range --make-bed --out Mayo_LDREF_Filt

rm -r Mayo/
#SNPS REMAINING COMPARED TO TOTAL LDREF SET:
##1,187,249 out of 1,190,321 (  99.74% )

#Rename Snps with rsIDs & Centimorgan Distance
python BimAnotator.py TotalLDRef.bim Mayo_LDREF_Filt.bim
mv Mayo_LDREF_FiltRenamed.bim Mayo_LDREF_Filt.bim

```

##Merge Chromsomes in Plink Format 
for i in {2..22} 
do 
    echo GWAS_Plink/MSSB_Chr_$i >> MergedGWAS.txt
done 
plink --threads $CORES --bfile GWAS_Plink/MSSB_Chr_1 --merge-list MergedGWAS.txt --make-bed --allow-no-sex --out GWASTotal
rm -r WGS_Plink/

#Merge  LDREF Data
for i in {2..22} 
do 
    echo LDREF/1000G.EUR.$i>> MergedChrs.txt
done 
plink --bfile LDREF/1000G.EUR.1 --threads $CORES --merge-list MergedChrs.txt --make-bed --allow-no-sex --out TotalLDRef

#Write SNP List for LDREF Data
#plink --threads $CORES --bfile INDV_1000G_Pops --write-snplist --out LDREF_SNPs

#Reformat SNP List for plink filter.txt
awk '{ print $1"\t"$4"\t"$4"\t"$2 }' 1000Genomes_Merged.bim > LDREF_SNPs.txt

##Filter for SNPs in 1000G - LDREF_FiltSNPs
plink --threads $CORES --bfile GWASTotal --extract LDREF_SNPs.txt --make-bed --allow-no-sex --out MSBB_LDREF_FiltSNPs

#Rename Snps with rsIDs & Centimorgan Distance
#python BimAnotator.py TotalLDRef.bim LDREF_FiltSNPs.bim
```

## ReName the MSBB SNPS so the can be combined
```{python ReNamer}
import sys
import os
import re

###This Script reannotates bim file name from a snp list
#Usage:
# python BimAnotator.py TotalLDRef.bim LDREF_FiltSNPs.bim

InFile = TotalLDRef.bim       #sys.argv[1];
#Gene_Annots/hg19/Merged_hg19_Filt_ENGS.bed
InFileb = LDREF_FiltSNPs.bim  #sys.argv[2];

RS_IDs = {}
RS_Pos = {}

F1 = open(InFile)
for line in F1:
	LINE = line.rstrip('\r\n')
	lst = LINE.split('\t')

	RS_IDs[ ''.join([lst[0], '_', lst[3]]) ] = lst[1]
	RS_Pos[ ''.join([lst[0], '_', lst[3]]) ] = lst[2]
F1.close()

#Open Output
OUT = open( ''.join([ InFileb.replace(".bim", ""), 'Renamed.bim']), 'w' )

F1 = open(InFileb)
for line in F1:
	LINE = line.rstrip('\r\n')
	lst = LINE.split('\t')

	Name = RS_IDs[ ''.join([ lst[0], "_", lst[3] ])]
	Post = RS_Pos[ ''.join([ lst[0], "_", lst[3] ])]

	Entry = '\t'.join([ lst[0], Name, Post, lst[3], lst[4], lst[5] ])
	print >>OUT, Entry

F1.close()
```