---
title: "MSSM-Penn-Pitt - ACC and DLPFC - FEATURECOUNTS - GRCh38 - CQN - Iterative Model"
author: "Addapted from Kelsey Montgomery"
date: "11/8/2019"
output: html_document
---

Date of analysis update: "`r date()`"

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, cache=FALSE}
##It is assumed your working directory is "~/ampad-DiffExp/gene_level_analysis"
## Load required libraries
library(CovariateAnalysis) # get the package from devtools::install_github('th1vairam/CovariateAnalysis@dev')
library(data.table)
library(plyr)
library(tidyverse)
library(psych)
library(stringr)
library(limma)
library(edgeR)
library(biomaRt)
library(RColorBrewer)
library(cqn)
library(glmnet)
library(githubr) # get the package from devtools::install_github('brian-bot/githubr')
library(knitr)
library(doParallel)
library(foreach)
library(sva)
#library(reticulate)
library(R.utils)

cl = makeCluster(detectCores()-2)
registerDoParallel(cl)
options(xtable.type="html")
# Source modified utility functions from limma and sva package
source('~/TWAS/utilityFunctions/parallelDuplicateCorrelation.R')
source('~/TWAS/utilityFunctions/irwsva.build.R')
source('~/TWAS/utilityFunctions/f.pvalue.R')
knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

### Problem Genes in the iterative normalization:
```{r prob.genes, cache=TRUE, include=FALSE}

ProbGenes <- c('ENSG00000227232', 'ENSG00000248527', 'ENSG00000285812', 'ENSG00000272455', 'ENSG00000215915', 'ENSG00000269737', 'ENSG00000157911', 'ENSG00000226029', 'ENSG00000121775', 'ENSG00000225828', 'ENSG00000116525', 'ENSG00000197982', 'ENSG00000285728', 'ENSG00000225767', 'ENSG00000229359', 'ENSG00000270629', 'ENSG00000160679', 'ENSG00000177954', 'ENSG00000227741', 'ENSG00000143257', 'ENSG00000236535', 'ENSG00000118194', 'ENSG00000143862', 'ENSG00000240219', 'ENSG00000123838', 'ENSG00000272645', 'ENSG00000143643', 'ENSG00000223694', 'ENSG00000272865', 'ENSG00000286216', 'ENSG00000179270', 'ENSG00000235760', 'ENSG00000144031', 'ENSG00000115282', 'ENSG00000232931', 'ENSG00000236440', 'ENSG00000227331', 'ENSG00000272211', 'ENSG00000229056', 'ENSG00000144410', 'ENSG00000144452', 'ENSG00000152056', 'ENSG00000183977', 'ENSG00000230409', 'ENSG00000172940', 'ENSG00000232354', 'ENSG00000281691', 'ENSG00000162267', 'ENSG00000189283', 'ENSG00000239926', 'ENSG00000239997', 'ENSG00000279349', 'ENSG00000285943', 'ENSG00000241278', 'ENSG00000240695', 'ENSG00000114113', 'ENSG00000240875', 'ENSG00000206120', 'ENSG00000229638', 'ENSG00000225473', 'ENSG00000230732', 'ENSG00000242686', 'ENSG00000249592', 'ENSG00000013810', 'ENSG00000284727', 'ENSG00000249212', 'ENSG00000251022', 'ENSG00000250361', 'ENSG00000250910', 'ENSG00000164105', 'ENSG00000283355', 'ENSG00000261360', 'ENSG00000274441', 'ENSG00000152669', 'ENSG00000205571', 'ENSG00000251634', 'ENSG00000171530', 'ENSG00000061455', 'ENSG00000164406', 'ENSG00000285605', 'ENSG00000241112', 'ENSG00000285868', 'ENSG00000241956', 'ENSG00000124588', 'ENSG00000189134', 'ENSG00000206341', 'ENSG00000204227', 'ENSG00000156639', 'ENSG00000272137', 'ENSG00000146242', 'ENSG00000255330', 'ENSG00000118514', 'ENSG00000146576', 'ENSG00000235859', 'ENSG00000183696', 'ENSG00000228735', 'ENSG00000234500', 'ENSG00000274080', 'ENSG00000205485', 'ENSG00000121716', 'ENSG00000271664', 'ENSG00000260555', 'ENSG00000236018', 'ENSG00000260721', 'ENSG00000249258', 'ENSG00000158815', 'ENSG00000197181', 'ENSG00000253200', 'ENSG00000134013', 'ENSG00000250979', 'ENSG00000253948', 'ENSG00000169946', 'ENSG00000253372', 'ENSG00000104529', 'ENSG00000234740', 'ENSG00000164970', 'ENSG00000237159', 'ENSG00000274026', 'ENSG00000136925', 'ENSG00000237461', 'ENSG00000196739', 'ENSG00000182752', 'ENSG00000286196', 'ENSG00000234771', 'ENSG00000280474', 'ENSG00000214402', 'ENSG00000197070', 'ENSG00000205696', 'ENSG00000285852', 'ENSG00000285737', 'ENSG00000188234', 'ENSG00000204172', 'ENSG00000178440', 'ENSG00000166343', 'ENSG00000229990', 'ENSG00000236552', 'ENSG00000155287', 'ENSG00000107819', 'ENSG00000273521', 'ENSG00000203772', 'ENSG00000177595', 'ENSG00000117983', 'ENSG00000247595', 'ENSG00000254622', 'ENSG00000162236', 'ENSG00000213465', 'ENSG00000110025', 'ENSG00000131620', 'ENSG00000241170', 'ENSG00000186174', 'ENSG00000186166', 'ENSG00000240970', 'ENSG00000254428', 'ENSG00000254909', 'ENSG00000110811', 'ENSG00000111664', 'ENSG00000111678', 'ENSG00000205885', 'ENSG00000256589', 'ENSG00000256682', 'ENSG00000247903', 'ENSG00000257433', 'ENSG00000139631', 'ENSG00000185482', 'ENSG00000123329', 'ENSG00000166987', 'ENSG00000178498', 'ENSG00000224713', 'ENSG00000213363', 'ENSG00000110887', 'ENSG00000257097', 'ENSG00000251497', 'ENSG00000279500', 'ENSG00000279283', 'ENSG00000198553', 'ENSG00000043355', 'ENSG00000126216', 'ENSG00000233695', 'ENSG00000258515', 'ENSG00000270433', 'ENSG00000276698', 'ENSG00000257986', 'ENSG00000258648', 'ENSG00000258682', 'ENSG00000196553', 'ENSG00000100632', 'ENSG00000258376', 'ENSG00000258891', 'ENSG00000183648', 'ENSG00000202191', 'ENSG00000223403', 'ENSG00000246451', 'ENSG00000201992', 'ENSG00000215252', 'ENSG00000243789', 'ENSG00000242866', 'ENSG00000250130', 'ENSG00000259619', 'ENSG00000247982', 'ENSG00000255346', 'ENSG00000246877', 'ENSG00000169752', 'ENSG00000274376', 'ENSG00000255769', 'ENSG00000249487', 'ENSG00000261505', 'ENSG00000261123', 'ENSG00000259895', 'ENSG00000279520', 'ENSG00000262312', 'ENSG00000188897', 'ENSG00000243716', 'ENSG00000166851', 'ENSG00000224310')
               
 tmp<-c(              'ENSG00000262155', 'ENSG00000260853', 'ENSG00000279789', 'ENSG00000156858', 'ENSG00000261439', 'ENSG00000089505', 'ENSG00000280334', 'ENSG00000280214', 'ENSG00000196436', 'ENSG00000140905', 'ENSG00000280182', 'ENSG00000226180', 'ENSG00000224888', 'ENSG00000170296', 'ENSG00000263934', 'ENSG00000266385', 'ENSG00000073605', 'ENSG00000236194', 'ENSG00000266967', 'ENSG00000262265', 'ENSG00000121104', 'ENSG00000243678', 'ENSG00000263089', 'ENSG00000263499', 'ENSG00000213246', 'ENSG00000172809', 'ENSG00000279382', 'ENSG00000262580', 'ENSG00000235552', 'ENSG00000263847', 'ENSG00000278983', 'ENSG00000278971', 'ENSG00000273218', 'ENSG00000268189', 'ENSG00000279203', 'ENSG00000279716', 'ENSG00000280332', 'ENSG00000267419', 'ENSG00000105185', 'ENSG00000239382', 'ENSG00000267053', 'ENSG00000233527', 'ENSG00000267254', 'ENSG00000171777', 'ENSG00000105221', 'ENSG00000268460', 'ENSG00000063177', 'ENSG00000104852', 'ENSG00000074219', 'ENSG00000167608', 'ENSG00000267838', 'ENSG00000177732', 'ENSG00000088832', 'ENSG00000131044', 'ENSG00000101421', 'ENSG00000222460', 'ENSG00000183798', 'ENSG00000101104', 'ENSG00000124172', 'ENSG00000286018', 'ENSG00000160224', 'ENSG00000223901', 'ENSG00000239415', 'ENSG00000234409', 'ENSG00000250479', 'ENSG00000279738', 'ENSG00000205702', 'ENSG00000226450', 'ENSG00000182841', 'ENSG00000279175', 'ENSG00000280434', 'ENSG00000235347', 'ENSG00000170638', 'ENSG00000185448', 'ENSG00000271199', 'ENSG00000198960', 'ENSG00000286000', 'ENSG00000241489', 'ENSG00000213397', 'ENSG00000198938', 'ENSG00000137970', 'ENSG00000230470', 'ENSG00000234781', 'ENSG00000233204', 'ENSG00000214389', 'ENSG00000235631', 'ENSG00000280379', 'ENSG00000279628', 'ENSG00000276663', 'ENSG00000263551', 'ENSG00000265316', 'ENSG00000231875', 'ENSG00000235790', 'ENSG00000242516', 'ENSG00000272040', 'ENSG00000250472', 'ENSG00000281706', 'ENSG00000184616', 'ENSG00000285581', 'ENSG00000238184', 'ENSG00000258136', 'ENSG00000136378', 'ENSG00000140682', 'ENSG00000279251', 'ENSG00000108375', 'ENSG00000279917', 'ENSG00000269318', 'ENSG00000270804', 'ENSG00000274322', 'ENSG00000235958', 'ENSG00000235038', 'ENSG00000228686', 'ENSG00000058673', 'ENSG00000234601', 'ENSG00000236549', 'ENSG00000091181', 'ENSG00000145020', 'ENSG00000248734', 'ENSG00000277769', 'ENSG00000146285', 'ENSG00000232131', 'ENSG00000228495', 'ENSG00000231760', 'ENSG00000272076', 'ENSG00000168828', 'ENSG00000279822', 'ENSG00000235010', 'ENSG00000250397', 'ENSG00000255159', 'ENSG00000258102', 'ENSG00000286067', 'ENSG00000274317', 'ENSG00000180662', 'ENSG00000259499', 'ENSG00000261207', 'ENSG00000175311', 'ENSG00000260751', 'ENSG00000261291', 'ENSG00000278668', 'ENSG00000266598', 'ENSG00000280248', 'ENSG00000276952', 'ENSG00000277829', 'ENSG00000223799', 'ENSG00000213790', 'ENSG00000271916', 'ENSG00000249863', 'ENSG00000267655', 'ENSG00000133872', 'ENSG00000224689','ENSG00000255389', 'ENSG00000253230', 'ENSG00000247137', 'ENSG00000170906', 'ENSG00000182195' )

 
 ProbGenes <- c(ProbGenes,tmp)
```

### Data download
#### Obtain count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
downloadFile <- function(id){
  fread(syn_temp$get(id)$path, data.table = F)
}
downloadFile_version <- function(id , version){
  fread(syn_temp$get(id, version = version)$path, data.table = F)
}
# Download reprocessed counts
counts = 'syn21071622'
ALL_USED_IDs = counts
counts = downloadFile(counts) %>% data.frame()
counts$Chr <- NULL
counts$Start <- NULL
counts$End <- NULL
counts$Strand <- NULL
# Parse gene lengths
gene_lengths = counts[,c("Geneid", "Length")]
counts$Length <- NULL
gene_lengths <- rename(gene_lengths, gene_id = Geneid)
# Get ancestry vector calculated using gemtools
ANCESTRY_ID = 'syn2511399'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = ANCESTRY_ID
ANCESTRY = downloadFile(ANCESTRY_ID) %>% 
  plyr::rename(c('DNA_report..Genotyping.Sample_ID' = 'SNP_report:Genotyping_Sample_ID'))
# Get genotype ids from synapse
GENOTYPE_ID = 'syn16816490'
ALL_USED_IDs = c(ALL_USED_IDs, GENOTYPE_ID)
GENOTYPE = downloadFile(GENOTYPE_ID) %>% 
  dplyr::select(Individual_ID, `SNP_report:Genotyping_Sample_ID`) %>%
  dplyr::inner_join(ANCESTRY)
#Get Metadata
metadata = 'syn16816488' 
ALL_USED_IDs[length(ALL_USED_IDs)+1] = metadata
metadata = downloadFile_version(metadata, version = 13)  %>% 
  dplyr::select(Individual_ID, Institution, Reported_Gender, Sex, Ethnicity, Age_of_Death, `PMI_(in_hours)`, Dx,
                Sample_RNA_ID, one_of('rnaSeq_isolation:RIN',
                                      'rnaSeq_report:Ribozero_Batch', 'rnaSeq_report:Library_Batch', 
                                      'rnaSeq_report:Flowcell_Batch','rnaSeq_dissection:Brain_Region',
                                      'rnaSeq_report:Exclude?',
                                      'rnaSeq_report:Mapped_Reads',
                                      'rnaSeq_report:Intragenic_Rate','rnaSeq_report:Intronic_Rate',
                                      'rnaSeq_report:Intergenic_Rate','rnaSeq_report:Genes_Detected',
                                      'rnaSeq_report:Expression_Profiling_Efficiency',
                                      'rnaSeq_report:rRNA_Rate',
                                      'rnaSeq_report:Total_Reads', 'rnaSeq_report:Percent_Aligned'))
METADATA = metadata %>%
  dplyr::left_join(GENOTYPE) %>%
  dplyr::rename(PMI = `PMI_(in_hours)`,
                Sample_ID = Sample_RNA_ID,
                ReportExclude = `rnaSeq_report:Exclude?`,
                SampleID = Sample_RNA_ID,
                LibraryBatch = `rnaSeq_report:Library_Batch`,
                FlowcellBatch = `rnaSeq_report:Flowcell_Batch`,
                RibozeroBatch = `rnaSeq_report:Ribozero_Batch`,
                Tissue = `rnaSeq_dissection:Brain_Region`,
                MappedReads = `rnaSeq_report:Mapped_Reads`,
                IntragenicRate = `rnaSeq_report:Intragenic_Rate`, 
                IntronicRate = `rnaSeq_report:Intronic_Rate`, 
                IntergenicRate = `rnaSeq_report:Intergenic_Rate`, 
                GenesDetected = `rnaSeq_report:Genes_Detected`,
                ExpProfEfficiency = `rnaSeq_report:Expression_Profiling_Efficiency`, 
                rRNARate = `rnaSeq_report:rRNA_Rate`,
                TotalReads = `rnaSeq_report:Total_Reads`, 
                AlignmentRate = `rnaSeq_report:Percent_Aligned`,
                RIN = `rnaSeq_isolation:RIN`) %>%
  dplyr::select(SampleID, Individual_ID, Institution, Reported_Gender,Sex, Ethnicity, Age_of_Death, PMI, Dx, 
                RIN, ReportExclude, EV.1, EV.2, EV.3, EV.4, EV.5,
                LibraryBatch, FlowcellBatch, RibozeroBatch, Tissue, MappedReads, TotalReads, GenesDetected, AlignmentRate,
                IntragenicRate, IntergenicRate, IntronicRate, ExpProfEfficiency, rRNARate) %>% 
  dplyr::filter(Institution %in% c("MSSM", "Penn", "Pitt")) %>% 
  dplyr::filter(SampleID %in% colnames(counts), !is.na(SampleID)) %>%
  dplyr::mutate(Dx = forcats::fct_recode(Dx,  Other = "AFF", Other = "BP", Control = "Control", SCZ = "SCZ"))
```

### Data preprocessing
```{r preprocess.data}
ind = METADATA$SampleID[which(METADATA$ReportExclude == 1)]
writeLines(paste('Following',length(ind),'samples are marked exclude'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 
ind = METADATA$SampleID [is.na(METADATA$Ethnicity) | is.na(METADATA$Institution) | is.na(METADATA$Dx)]
writeLines(paste('Following', length(ind), 'counts are missing clinical metadata'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 
ind = METADATA$SampleID [is.na(METADATA$PMI)]
writeLines(paste('Following', length(ind), 'counts are missing PMI'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 
ind = METADATA$SampleID [is.na(METADATA$Reported_Gender)]
writeLines(paste('Following', length(ind), 'counts are missing gender'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 
ind = METADATA$SampleID [is.na(METADATA$Age_of_Death)]
writeLines(paste('Following', length(ind), 'counts are missing age of death'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind))
ind = METADATA$SampleID [is.na(METADATA$EV.1)]
writeLines(paste('Following', length(ind), 'counts are missing ancestry information'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind))
```

```{r preprocess.data1, results='asis'}
# Match covariates to expression data
indToRetain = intersect(METADATA$SampleID, colnames(counts))
rownames(counts) <- counts$Geneid
counts = counts[,indToRetain]
rownames(METADATA) = METADATA$SampleID
METADATA = METADATA[indToRetain,]
METADATA %>% 
  group_by(Dx, Tissue) %>% 
  summarise(count = n()) %>% 
  spread(Tissue, count) %>%
  kable()
#_# Add in Dx.TissueSource
METADATA = METADATA %>%
  dplyr::mutate(Dx.Tissue = paste(Dx, Tissue, sep = '.')) %>%
  dplyr::mutate(Dx.TissueSource = paste(Dx, Tissue, Institution, sep = '.'))
```
Following sample are removed `r paste(removedIDs, collapse = ',')`

```{r gene.param}
## Get GC content from biomart
backgroundGenes = data.frame(gene_id = rownames(counts)) %>%
  dplyr::mutate(id = gene_id) %>%
  tidyr::separate(id, c('ensembl_gene_id','position'), sep = '\\.')
# Define biomart object
mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", host = "jul2019.archive.ensembl.org", dataset = "hsapiens_gene_ensembl")
# Query biomart
Ensemble2HGNC <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol", "percentage_gene_gc_content", "gene_biotype", "chromosome_name"),
                       filters = "ensembl_gene_id", values = backgroundGenes$ensembl_gene_id,
                       mart = mart)
GENE.GC.CONT = Ensemble2HGNC %>%
  dplyr::left_join(backgroundGenes) %>% 
  dplyr::select(gene_id, percentage_gene_gc_content, chromosome_name) %>%
  unique
rownames(GENE.GC.CONT) = GENE.GC.CONT$gene_id
```

### Sex Check
```{r Reported_Gender_plots}
counts$gene_id = rownames(counts) 
REPORTED.GENDER.COUNTS = GENE.GC.CONT %>% 
  left_join(counts) %>%
  dplyr::select(-one_of("percentage_gene_gc_content")) %>%
  filter(chromosome_name == "X" |chromosome_name == "Y") %>% 
  tidyr::gather(key = item, value = value, -c(gene_id, chromosome_name)) %>%
  mutate(value = log(value)) %>%
  rename(`counts(log)`= value) %>% 
  rename(SampleID = item) %>%
  left_join(METADATA[,c("SampleID", "Reported_Gender")]) %>% 
  rename(`Reported Gender` = Reported_Gender) 
##XIST and UTY expression 
#ENSG00000229807.10 and ENSG00000183878.15 
#ENSG00000229807.11
FILT <- REPORTED.GENDER.COUNTS %>% 
  filter(gene_id == "ENSG00000229807.11" | gene_id == "ENSG00000183878.15") %>% 
  dplyr::select(-one_of("chromosome_name")) %>% 
  tidyr::spread(key = gene_id, value = `counts(log)`) %>% 
  mutate(XIST = as.numeric(`ENSG00000229807.11`)) %>% 
  mutate(UTY = as.numeric(`ENSG00000183878.15`))
p = ggplot(FILT, aes (x= XIST, y = UTY)) 
p = p + geom_point(aes(color=`Reported Gender`)) 
p
```

### Covariate clustering
Determine relationship between covariates
```{r covariates.clustering}
#primaryVariable <- c("Tissue.SourceDiagnosis", "Tissue.Diagnosis", "Tissue.APOE4")
primaryVariable <- c("Dx.TissueSource", "Dx.Tissue")
#_# ADD DIAGNOSIS HERE TO REGRESS FOR TWAS
FactorCovariates <- c('Individual_ID','FlowcellBatch', "Institution", "Reported_Gender", "LibraryBatch", "Tissue", "Dx.Tissue", "Dx", "RibozeroBatch")
#_#FactorCovariates <- c('Individual_ID', "Institution","Dx.Tissue", "Reported_Gender", "Dx", "RibozeroBatch")
ContCovariates <- c("Age_of_Death", "PMI", "RIN", "MappedReads", "IntragenicRate", "IntronicRate", "IntergenicRate", "GenesDetected", "ExpProfEfficiency", "rRNARate", "TotalReads","AlignmentRate", "EV.1", "EV.2", "EV.3", "EV.4", "EV.5")

#METADATA <- METADATA[ METADATA$Tissue == 'DLPFC', ]
#counts <- counts[ , METADATA$SampleID]

# Find inter relation between factor covariates
METADATA <- METADATA[ !duplicated(METADATA), ]
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates),drop=F]
rownames(COVARIATES) <- METADATA$ID
# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.character)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)
```
Correlation/association between covariates at an FDR <= 0.1
```{r covariates.correlation, cache=FALSE, fig.width=10, fig.height=6}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES, PVAL = 0.05)
draw(COVARIATES.CORRELATION$plot, heatmap_legend_side = 'left', padding  = unit(c(18,2,2,18), 'mm'))
```
### Explore metadata
```{r data.explore, cache=FALSE, fig.width = 12, fig.height = 8}

METADATA <- METADATA[ METADATA$Tissue == 'DLPFC', ]
counts <- counts[ , METADATA$SampleID]
COVARIATES=COVARIATES[ COVARIATES$Tissue == 'DLPFC' ,]
my.theme = theme(legend.position = 'top', axis.text.x = element_text(angle = 90, hjust = 1))
# RIN
p = list()
p[[1]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = RIN)) + geom_boxplot()
p[[1]] = p[[1]] + ggtitle('RIN') + my.theme
# AgeAtDeath
p[[2]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = as.numeric(Age_of_Death))) + geom_boxplot()
p[[2]] = p[[2]] + ggtitle('Age_of_Death') + my.theme
# PMI
p[[3]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = as.numeric(PMI))) + geom_boxplot()
p[[3]] = p[[3]] + ggtitle('PMI') + my.theme
# Ribosomal bases
p[[4]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = as.numeric(rRNARate))) + geom_boxplot()
p[[4]] = p[[4]] + ggtitle('Fraction of Ribosomal Bases') + my.theme
# Intronic bases
p[[5]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = as.numeric(IntronicRate))) + geom_boxplot()
p[[5]] = p[[5]] + ggtitle('Fraction Intronic Bases') + my.theme
# Intergenic bases
p[[6]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = as.numeric(IntergenicRate))) + geom_boxplot()
p[[6]] = p[[6]] + ggtitle('Fraction Intergenic Bases') + my.theme
multiplot(plotlist = p, cols = 3)
```

### Filter genes
* Remove genes that have less than 1 cpm counts in at least 50% of samples per Tissue x Diagnosis
* Remove genes with missing gene length and percentage GC content
```{r filter.genes}
COUNT<-counts[,METADATA$SampleID]
row.names(COVARIATES) <- METADATA$SampleID

Unis <- do.call(rbind, strsplit(as.character(row.names(COUNT)), '\\.'))[,1]
dups <- Unis[ duplicated(Unis) ]
Prosps <- do.call(rbind, strsplit(as.character(row.names(COUNT)), '\\.'))
Prosps <- Prosps[ (Prosps[,1] %in% dups) == F,]
keeps <- paste0( Prosps[,1], '.', Prosps[,2] )

COUNT <- COUNT[ keeps, ]

row.names(COUNT) <- do.call(rbind, strsplit(as.character(row.names(COUNT)), '\\.'))[,1]

#colnames(COVARIATES)[1] <- 'SampleID'
#colnames(COUNT) <- paste0( "CMC_", gsub("RNA_BP_PFC_","",colnames(COUNT)) )
#colnames(COUNT) <- gsub("RNA_PFC_","",colnames(COUNT)) 
#colnames(COUNT) <- gsub("RNA_ACC_","",colnames(COUNT)) 
#colnames(COUNT) <- gsub("BP_","",colnames(COUNT)) 
#tmp <- do.call(rbind, strsplit(colnames(COUNT), '_'))
#tmp[,3] <- str_pad(tmp[,3], 3, pad = "0")
#colnames(COUNT) <- paste0(tmp[,1], '_', tmp[,2], '_',tmp[,3])
#COVARIATES$SampleID <- as.character(COVARIATES$SampleID)
#row.names(COVARIATES) <- COVARIATES$SampleID
#table(METADATA$SampleID %in% colnames(COUNT))
#table(colnames(COUNT) %in% METADATA$SampleID )

genesToAnalyze = COVARIATES %>%
  rownameToFirstColumn('SampleID') %>%
  dlply(.(Dx.Tissue), .fun = function(mtd, count){
    processed.counts = getGeneFilteredGeneExprMatrix(count[,mtd$SampleID],
                                                     MIN_GENE_CPM=1, 
                                                     MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.5)
    
    #processed.counts = getGeneFilteredGeneExprMatrix(COUNT[,METADATA$SampleID],
                                                     #MIN_GENE_CPM=1, 
                                                     #MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.5)
    processed.counts$filteredExprMatrix$genes
  }, COUNT)


GENE.PARAM = syn_temp$get('syn21156161', version = 3)$path %>% 
  read.table(sep = '\t', header = T, stringsAsFactors = F)
#row.names(GENE.PARAM) <- GENE.PARAM$ensembl_gene_id
GENE.LEN = dplyr::select(GENE.PARAM, ensembl_gene_id, gene.length) %>% unique()
rownames(GENE.LEN) = GENE.LEN$ensembl_gene_id
GENE.GC = dplyr::select(GENE.PARAM, ensembl_gene_id, percentage_gc_content) %>% unique()
rownames(GENE.GC) = GENE.GC$ensembl_gene_id

genesToAnalyze = unlist(genesToAnalyze) %>% 
  unique() %>%
  intersect(GENE.GC$ensembl_gene_id[!is.na(GENE.GC$percentage_gc_content)]) %>%
  intersect(GENE.LEN$ensembl_gene_id[!is.na(GENE.LEN$gene.length)]) %>%
  setdiff(c("N_unmapped", "N_multimapping", "N_noFeature", "N_ambiguous"))

PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[genesToAnalyze, ], 
                                                 MIN_GENE_CPM=0, 
                                                 MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
# Check gene biotype
## Define biomart object
mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", host = "jul2019.archive.ensembl.org", dataset = "hsapiens_gene_ensembl")
## Query biomart
Ensemble2HGNC <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol", "gene_biotype"),
                       filters = "ensembl_gene_id", 
                       values = PROCESSED_COUNTS$filteredExprMatrix$genes,
                       mart = mart)
summary(factor(Ensemble2HGNC$gene_biotype)) %>%
  rownameToFirstColumn('Biotype') %>%
  dplyr::rename(fraction = DF) %>%
  dplyr::mutate(fraction = fraction/dim(PROCESSED_COUNTS$filteredExprMatrix$genes)[1]) %>%
  dplyr::filter(fraction >= 0.01) %>%
  kable
```
Processing `r dim(PROCESSED_COUNTS$filteredExprMatrix)[1]` genes in `r dim(PROCESSED_COUNTS$filteredExprMatrix)[2]` samples


### Library Normalisation
Library normalisation is performed using cqn (conditional quantile normalisation)
```{r cqn}
row.names(GENE.GC) <- GENE.GC$ensembl_gene_id
row.names(GENE.LEN) <- GENE.LEN$ensembl_gene_id
# Compute offset for gene length and gc content
CQN.GENE_EXPRESSION = cqn(PROCESSED_COUNTS$filteredExprMatrix$counts, 
                          x = GENE.GC[PROCESSED_COUNTS$filteredExprMatrix$genes$genes, 'percentage_gc_content'],
                          lengths = GENE.LEN[PROCESSED_COUNTS$filteredExprMatrix$genes$genes, 'gene.length'],
                          lengthMethod = "smooth", 
                          verbose = FALSE)
CQN.GENE_EXPRESSION$E = CQN.GENE_EXPRESSION$y + CQN.GENE_EXPRESSION$offset
```

### Outlier Analysis
#### Sample outliers
Outlier analysis is performed before library normalisation with raw cpm counts
```{r outlier.analysis, cache = FALSE, fig.width = 10}
#indToRemove = c("11311_CER", "1923_CER", "1923_TCX", "11396_TCX", "11294_TCX", "11408_TCX",
#                '1950_TCX', '1950_CER', '1925_TCX', '1957_CER')
# Find principal components of expression to plot
PC <- prcomp(CQN.GENE_EXPRESSION$E, scale.=T, center = T)
# Plot first 2 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])
plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID')) %>%
  dplyr::mutate(label = SampleID) %>%
  tidyr::separate(Dx.Tissue, c('Diagnosis', 'Tissue'), sep = '\\.')
#plotdata$label[!(plotdata$SampleID %in% indToRemove)] = ''
p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(shape=Institution, color=Diagnosis, size=RIN))
p <- p + theme_bw() + theme(legend.position="top") 
p <- p + facet_grid(Tissue~.+Diagnosis, scales = 'free')
p <- p + geom_text(aes(label= label), size=4, hjust=0)
p
# Plot abberent distribution of logcpm counts
tmp1 = CQN.GENE_EXPRESSION$E %>%
  rownameToFirstColumn('Gene.ID') %>%
  tidyr::gather(SampleID, logCPM, -Gene.ID) %>%
  left_join(COVARIATES %>%
              rownameToFirstColumn('SampleID')) %>%
  tidyr::separate(Dx.Tissue, c('Tissue', 'Diagnosis'), sep = '\\.')
indToRemove <- ''
p = ggplot(tmp1 %>%
             #dplyr::filter(SampleID %in% indToRemove) %>%
             droplevels(),
           aes(x = logCPM)) + geom_density() 
p = p + theme(legend.position = 'top') + facet_grid(Tissue~.+Diagnosis)
p
indToRetain = setdiff(colnames(PROCESSED_COUNTS$filteredExprMatrix$counts), indToRemove)
PROCESSED_COUNTS$filteredExprMatrix$counts = PROCESSED_COUNTS$filteredExprMatrix$counts[,indToRetain]
CQN.GENE_EXPRESSION$E = CQN.GENE_EXPRESSION$E[,indToRetain]
COVARIATES = COVARIATES[indToRetain,]
tmp = COVARIATES %>%
  dplyr::group_by(Dx.Tissue) %>%
  dplyr::summarise(count = n()) %>%
  tidyr::separate(Dx.Tissue, c('Diagnosis', 'Tissue'), sep = '\\.') %>%
  tidyr::spread(Diagnosis, count)
```
Processing `r dim(PROCESSED_COUNTS$filteredExprMatrix)[1]` genes in `r dim(PROCESSED_COUNTS$filteredExprMatrix)[2]` samples

Based on the expression pattern following samples were tagged as outliers: `r paste(indToRemove, collapse = ', ')`

Distribution of samples are: `r kable(tmp)`


#### Gene outliers
Assign NA values to genes that are above and below 3 std deviation of its distribution
```{r winsorise.data}
# Set gene counts in specific samples that are deviating 3 sd from other samples to 3SD limit
LOG.CPM = apply(CQN.GENE_EXPRESSION$E, 1, function(x){
  mn = mean(x, na.rm = T)
  std.dev = sd(x, na.rm = T)
  
  x[x < (mn-3*std.dev)] = NA
  x[x > (mn+3*std.dev)] = NA
  return(x)
}) %>% t
CQN.GENE_EXPRESSION$E = LOG.CPM[ (row.names(LOG.CPM) %in% ProbGenes) == F, ]
CQN.GENE_EXPRESSION$E.no.na = CQN.GENE_EXPRESSION$E
CQN.GENE_EXPRESSION$E.no.na[is.na(CQN.GENE_EXPRESSION$E.no.na)] = 0
LIB.SIZE = colSums(PROCESSED_COUNTS$filteredExprMatrix$counts)
NEW.COUNTS = (2^LOG.CPM) * t(replicate(dim(LOG.CPM)[1], LIB.SIZE))/1e6

dim(NEW.COUNTS)
NEW.COUNTS <- NEW.COUNTS[ (row.names(NEW.COUNTS) %in% ProbGenes) == F, ]
dim(NEW.COUNTS)
```


### Sample clustering
PCA based clustering of samples
```{r decompse.normalise.data, cache=FALSE, fig.height=8, fig.width=8, results='asis', cache=FALSE}
# Find principal components of expression to plot
PC <- prcomp(CQN.GENE_EXPRESSION$E.no.na, scale.=T, center = T)
# Plot first 2 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])
plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID')) %>%
  tidyr::separate(Dx.Tissue, c( 'Diagnosis', 'Tissue'), sep = '\\.')
p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(shape=Institution, color=Diagnosis, size=RIN))
p <- p + theme_bw() + theme(legend.position="top") + facet_grid(Tissue+.~Diagnosis, scales = 'free_y')
p
```


Tree based clustering of samples
```{r decompse.normalise.data.1, cache=FALSE, fig.height=6, fig.width=10, results='asis'}
COVARIATES.tmp = (COVARIATES[,c(FactorCovariates), drop = F])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0
tree = hclust(as.dist(t(CQN.GENE_EXPRESSION$E.no.na)))
for( i in 1:dim(COVARIATES.tmp)[2]){
  COVARIATES.tmp[,i] <- as.numeric(COVARIATES.tmp[,i])
}
COVARIATES.tmp <- COVARIATES.tmp[,2:dim(COVARIATES.tmp)[2]]
cols = WGCNA::labels2colors(COVARIATES.tmp);
WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp))
```

### Distribution of samples (log cpm)
```{r lcpm.dist, cache=FALSE, fig.height=10, fig.width=20}
# Plot abberent distribution of logcpm counts
tmp1 = CQN.GENE_EXPRESSION$E %>%
  rownameToFirstColumn('Gene.ID') %>%
  tidyr::gather(SampleID, logCPM, -Gene.ID) %>%
  left_join(COVARIATES %>%
              rownameToFirstColumn('SampleID')) %>%
  tidyr::separate(Dx.Tissue, c( 'Diagnosis', 'Tissue' ), sep = '\\.')
p = ggplot(tmp1, aes(x = logCPM)) + geom_density() 
p = p + theme(legend.position = 'NONE') + facet_grid(Tissue+.~Diagnosis, scale = 'free')
p
```

Coexpression of genes 
```{r coexp1, cache=FALSE, fig.height=5, fig.width=5}
cr = cor(t(CQN.GENE_EXPRESSION$E.no.na))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```

### Significant Covariates
Correlation between pca of unadjusted mRNA expression and covariates are used to find significant covariates
```{r preadj.covariates, cache=TRUE}
# Find correlation between PC's of gene expression with covariates
#CQN.GENE_EXPRESSION$E.no.na <- as.data.frame(CQN.GENE_EXPRESSION$E.no.na)
COVARIATES1 <- COVARIATES[ , (colnames(COVARIATES) %in% 'Tissue') == F]
preAdjustedSigCovars = runPCAandPlotCorrelations(CQN.GENE_EXPRESSION$E.no.na, 
                                                 COVARIATES1,
                                                 'NULL design(voom-normalized)', 
                                                 isKeyPlot=TRUE, 
                                                 MIN_PVE_PCT_PC = 1)
```

Significant covariates to adjust at FDR 0.1 are `r preAdjustedSigCovars$significantCovars`
```{r preadj.covariates.plot, cache=FALSE, fig.width=12, fig.height=8}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```

### Normalisation (iterative design)
Since many covariates are correlated, re-normalising and re-adjusting COUNTS with an iterative design matrix.

NOTE:
1. Using a mixed effect model where random effect is chosen as Individual_ID
2. Adding Source, FLOWCELL and Gender a priori to variable selection
3. Primary variable of interest Tissue.Diagnosis is excluded from the pool of available covariates for selection
```{r iterative.norm, results='asis'}
# Primary variable of interest
postAdjustCovars = c('Institution', 'FlowcellBatch', 'Reported_Gender');
# Assign residual covariates
residualCovars = setdiff(preAdjustedSigCovars$significantCovars, 
                         c(postAdjustCovars, primaryVariable, 'Individual_ID'))
residualSigCovars = preAdjustedSigCovars
covariatesEffects = preAdjustedSigCovars$Effects.significantCovars[residualCovars]
postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects))) %>% unique()
notEstimable = c()

loopCount = 0 
while(length(residualSigCovars$significantCovars)!=0 && loopCount <= 20){
  writeLines(paste0('Working on Loop: ', loopCount))
  writeLines(paste('Using following covariates in the model:',
                   paste(postAdjustCovars, collapse=', '),
                   'as fixed effects and Individual_ID as random effect'))
  
  # Post adjusted design matrix
  DM1 = getDesignMatrix(COVARIATES1[,postAdjustCovars,drop=F],Intercept = F)
  DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]
  
  # Estimate voom weights for dispersion control
  cnts = NEW.COUNTS; cnts[is.na(cnts)] = 0
  VOOM.GENE_EXPRESSION = voom(cnts, design=DM1$design, 
                              plot=F, na.rm = T)
  
  # Calculate correlation values of random effects
  VOOM.GENE_EXPRESSION$E = CQN.GENE_EXPRESSION$E
  correlation = parallelDuplicateCorrelation(VOOM.GENE_EXPRESSION,
                                             block = COVARIATES1$Individual_ID,
                                             method = 'lmer')
  # correlation = list(); correlation$cor = 0.3819258
  
  if (!is.nan(correlation$cor)){
    # Re-estimate voom weights
    VOOM.GENE_EXPRESSION = voom(cnts, 
                                design=DM1$design, plot=F, 
                                block = COVARIATES1$Individual_ID, 
                                correlation = correlation$cor)
    
    # Fit linear model using new weights and new design
    VOOM.ADJUSTED.FIT = lmFit(CQN.GENE_EXPRESSION$E,
                              design = DM1$design,
                              weights = VOOM.GENE_EXPRESSION$weights,
                              block = COVARIATES1$Individual_ID, 
                              correlation = correlation$cor)
    
    # Residuals after normalisation
    RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(VOOM.ADJUSTED.FIT,
                                                  CQN.GENE_EXPRESSION$E)
  } else {
    notEstimable = c(notEstimable, postAdjustCovars[length(postAdjustCovars)])
    postAdjustCovars = postAdjustCovars[1:(length(postAdjustCovars)-1)]
  }
  
  # Residual covariates to choose from
  residCovars <- setdiff(c(FactorCovariates,ContCovariates),
                         c(postAdjustCovars, primaryVariable, 'Individual_ID', notEstimable))
  residCovars <- residCovars[ (residCovars == 'Tissue') == F ]
  
  # Find PC of residual gene expression and significant covariates that are highly correlated with PCs
  expr = RESIDUAL.GENE_EXPRESSION; expr[is.na(expr)] = 0
  
  residualSigCovars = runPCAandPlotCorrelations(expr, 
                                                COVARIATES1[, residCovars, drop=F], 
                                                'adjusted design(voom-normalized)',
                                                isKeyPlot=TRUE)
  
  # Add postadjusted covariates (if any)
  residCovars = setdiff(residualSigCovars$significantCovars, c(postAdjustCovars, primaryVariable, notEstimable))
  covariatesEffects = residualSigCovars$Effects.significantCovars[residCovars]
  
  postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects)))
  loopCount = loopCount + 1
}
modelStr <- paste(paste(gsub('_','\\\\_',postAdjustCovars), collapse=', '),
                  'as fixed effects')
tmp <- paste('Using following covariates in the final model:', modelStr)
```
`r tmp`

### Sanity check
```{r residual.adj, cache=FALSE, fig.width=12, fig.height=8}
# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
residualSigCovars = runPCAandPlotCorrelations(expr, 
                                              COVARIATES1,
                                              'adjusted design(voom-normalized)',
                                              isKeyPlot=TRUE)
residualSigCovars[["PC_res"]][[2]]$plotData
```


Coexpression of genes 
```{r coexp2, cache=FALSE, fig.height=5, fig.width=5}
cr = cor(t(expr))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```
PCA of residual data
```{r decompse.normalise.data2.1, cache=FALSE, fig.height=10, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(expr, scale.=T, center = T)
# Plot first 4 PCs
plotdata <- data.frame(Individual_ID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])
#COVARIATES$SampleID <- COVARIATES$Individual_ID
plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'Individual_ID')) %>%
  tidyr::separate(Dx.Tissue, c('Diagnosis','Region'), sep = '\\.') %>% 
  dplyr::mutate(Region = factor(Region),
                Institution = factor(Institution))
p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(shape=Institution, color=Diagnosis, size=RIN))
p <- p + theme_bw() + theme(legend.position="right") + facet_grid(Region~., scales = 'free_y')
p
```
Tree based clustering of residual data
```{r decompse.normalise.data2.2, cache=FALSE, fig.height=8, fig.width=12, results='asis'}
# Eucledian tree based analysis
COVARIATES.tmp = data.matrix(COVARIATES[,FactorCovariates])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0
tree = hclust(as.dist(t(expr)))
COVARIATES.tmp <- COVARIATES.tmp[,2:dim(COVARIATES.tmp)[2]]
cols = WGCNA::labels2colors(COVARIATES.tmp);
WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp))
```

### Adjust data with covariates for Network Analysis
Identified covariates are regressed out from the expression matrix for network analysis
```{r network.adjust}
# Get design matrix
DESIGN.NET = getDesignMatrix(COVARIATES[, postAdjustCovars, drop = F], Intercept = F)
DESIGN.NET = DESIGN.NET$design[,linColumnFinder(DESIGN.NET$design)$indepCols]
# Estimate voom weights for dispersion control
VOOM.NET.WEIGHTS = voom(cnts, design=DESIGN.NET, plot=F)
# Re-estimate voom weights
VOOM.NET.WEIGHTS = voom(cnts, 
                        design = DESIGN.NET, plot=F,
                        block = COVARIATES$Donor_ID, 
                        correlation = correlation$cor)
# Fit linear model using new weights and new design
VOOM.NET.WEIGHTS$E = CQN.GENE_EXPRESSION$E
VOOM.NET.FIT = lmFit(VOOM.NET.WEIGHTS$E,
                     weights = VOOM.NET.WEIGHTS$weights,
                     design = VOOM.NET.WEIGHTS$design,
                     block = COVARIATES$Donor_ID, 
                     correlation = correlation$cor)
# Residuals after normalisation
RESIDUAL.NET.GENE_EXPRESSION = residuals.MArrayLM(VOOM.NET.FIT,
                                                  CQN.GENE_EXPRESSION$E)
```

### SVA Adjustments for eQTL analysis
Conditioned on primary variable (Tissue.Diagnosis) and identified covariates estimate surrogate variables using SVA package
```{r sva.adjust}

FactorCovariates <- c("Individual_ID", "FlowcellBatch", "Institution", "Reported_Gender", "LibraryBatch", "Dx.Tissue", "Dx", "RibozeroBatch")
ContCovariates <- c("Age_of_Death", "PMI", "RIN", "MappedReads", "IntragenicRate", "IntronicRate", "IntergenicRate", "GenesDetected", "ExpProfEfficiency", "rRNARate", "TotalReads","AlignmentRate", "EV.1", "EV.2", "EV.3", "EV.4", "EV.5")

# Find inter relation between factor covariates
#METADATA <- METADATA[ !duplicated(METADATA), ]
#COVARIATES1 = METADATA[,c(FactorCovariates,ContCovariates),drop=F]
#rownames(COVARIATES1) <- METADATA$ID
# Convert factor covariates to factors
COVARIATES1[,FactorCovariates] = lapply(COVARIATES1[,FactorCovariates], factor)
COVARIATES1[,ContCovariates] = lapply(COVARIATES1[,ContCovariates], as.character)
COVARIATES1[,ContCovariates] = lapply(COVARIATES1[,ContCovariates], as.numeric)


# Get design matrix
DESIGN = getDesignMatrix(COVARIATES1[, c(primaryVariable[2], postAdjustCovars), drop = F], Intercept = F)

#_#DESIGN$design = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
# Estimate voom weights for dispersion control
VOOM.WEIGHTS = voom(cnts, design=DESIGN$design, plot=F)
# Re-estimate voom weights
VOOM.WEIGHTS = voom(cnts,
                    design=DESIGN$design, plot=F,
                    block = COVARIATES1$Donor_ID, 
                    correlation = correlation$cor)
# Fit linear model using new weights and new design
VOOM.WEIGHTS$E = CQN.GENE_EXPRESSION$E
FIT = lmFit(VOOM.WEIGHTS$E,
            weights = VOOM.WEIGHTS$weights,
            design = VOOM.WEIGHTS$design[,colnames(VOOM.NET.FIT)],
            block = COVARIATES$Donor_ID, 
            correlation = correlation$cor)

# Get (null) design matrix
MODEL0 = DESIGN$design[,-grep(primaryVariable[2],colnames(DESIGN$design))] # Get null model by removing variable of interest
MODEL0 = MODEL0[,linColumnFinder(MODEL0)$indepCols]
MODEL1 = DESIGN$design
MODEL1 = MODEL1[,linColumnFinder(MODEL1)$indepCols]
# Get residuals from differential expression model
RESIDUAL.SVA.GENE_EXPRESSION = residuals.MArrayLM(FIT, CQN.GENE_EXPRESSION$E)
# Compute actual variance of all principal components
expr = RESIDUAL.SVA.GENE_EXPRESSION
expr[is.na(expr)] = 0
tmp = svd(expr)
actual.var = tmp$d^2/sum(tmp$d^2)
# Compute permuted variance of all principal components
permuted.var = foreach(i = 1:20, .combine = rbind) %dopar% {
  tmp.residual = t(apply(expr, 1, sample, replace = FALSE))
  tmp = svd(tmp.residual)
  permuted.var = tmp$d^2/sum(tmp$d^2)
}
permuted.var = apply(permuted.var, 2, mean)
NUM.SV = sum(actual.var >= permuted.var[1])
# Plot variance components
var.comp = data.frame(component = 1:length(actual.var), 
                      residual = actual.var, 
                      permuted.residual = permuted.var) %>%
  tidyr::gather(data, value, -component)
p = ggplot(var.comp %>% filter(component <= round(NUM.SV*1.2)), aes(x = component, y = value, color = data)) + geom_point()
p = p + geom_hline(yintercept=permuted.var[1], linetype = 'dashed') + xlab('Singular Dimension Index')
p = p + geom_vline(xintercept=NUM.SV, linetype = 'dashed') + ylab('Fraction of Variance Explained')
p = p + theme(legend.position = c(0.8, 0.8), legend.title = element_blank())
p
# Estimate surrogate variables
SURR.VAR = irwsva.build(CQN.GENE_EXPRESSION$E.no.na,
                        MODEL1, MODEL0, n.sv = NUM.SV, 
                        B = 30, tol = 1e-20)$sv
colnames(SURR.VAR) = paste0('SV',1:dim(SURR.VAR)[2])
rownames(SURR.VAR) = rownames(MODEL1)
# Estimate voom weights for dispersion control
VOOM.SVA.WEIGHTS = voom(cnts, design=cbind(MODEL1, SURR.VAR), plot=F)
# Re-estimate voom weights
VOOM.SVA.WEIGHTS = voom(cnts, 
                        design = cbind(MODEL1, SURR.VAR), plot=F,
                        block = COVARIATES1$Donor_ID, 
                        correlation = correlation$cor)
# Fit linear model using new weights and new design
VOOM.SVA.WEIGHTS$E = CQN.GENE_EXPRESSION$E
VOOM.SVA.FIT = lmFit(VOOM.SVA.WEIGHTS$E,
                     weights = VOOM.SVA.WEIGHTS$weights,
                     design = VOOM.SVA.WEIGHTS$design,
                     block = COVARIATES1$Donor_ID, 
                     correlation = correlation$cor)
# Residuals after normalisation
RESIDUAL.SVA.GENE_EXPRESSION = residuals.MArrayLM(VOOM.SVA.FIT,
                                                  CQN.GENE_EXPRESSION$E)
# Add variable of interest back to the residuals
varsToAddIn = grep("Tissue.Diagnosis", colnames(MODEL1), value = T)
RESIDUAL.SVA.GENE_EXPRESSION = RESIDUAL.SVA.GENE_EXPRESSION + 
  VOOM.SVA.FIT$coefficients[,varsToAddIn] %*% t(VOOM.SVA.FIT$design[,varsToAddIn])
```







```{r synapse.parameters, include=FALSE, cache=TRUE}
parentId = 'syn21470292';
activityName = 'Covariate and Diagnosis Regression';
activityDescription = 'CMCCovariate analysis and Regrsison of AD Diagnosis of aligned effective counts with GRCh38 with CQN normalisation (DLPFC)';
thisFileName <- 'CMC_Exp_DataProcessing.Rmd'
# Github link
thisRepo <- githubr::getRepo(repository = "jgockley62/TWAS", ref="branch", refName='master')
thisFile <- githubr::getPermlink(repository = thisRepo, repositoryPath=paste0('code/',thisFileName))
```

### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
activityName = 'CMC Covariate and Diagnosis Regression';
activityDescription = 'CMC Covariate analysis and Regrssion of Diagnosis of aligned effective counts with GRCh38 with CQN normalisation CMC';
CODE <- syn_temp$store(synapseclient$Folder(name = "Iterative_DLPFC_Normalization", parentId = parentId))
#Set Used SynIDs For Provenance
Syns_Used <- c("")

# Set annotations
all.annotations = list(
  dataType = 'mRNA',
  dataSubType = 'geneExp',
  summaryLevel = 'gene',
  assay	 = 'RNAseq',
  tissueTypeAbrv	= c('DLPFC'), 
  study = 'CMC', 
  organism = 'HomoSapiens',
  consortium	= 'CMC',
  normalizationStatus	= TRUE,
  normalizationType	= 'CQN',
  rnaquantification = 'RSEM',
  genomeAssemblyID = 'GRCh38'
)
# Store SVA results
write.table(RESIDUAL.SVA.GENE_EXPRESSION, file = 'CMC_SVA_Diagnosis_Corrected.tsv', sep = '\t', quote=F)
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='CMC_SVA_Diagnosis_Corrected.tsv', name = 'Normalised, covariates, surrogate and Diagnosis variable adjusted residual expression (for eQTL)', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataSubType = 'residualGeneExpForeQTL'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)
  
# Store residual gene expression for network analysis
RESIDUAL.NET.GENE_EXPRESSION %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'CMC_Diagnosis_Regressed_netResidualExpression.tsv', sep = '\t', row.names=F, quote=F)
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='CMC_Diagnosis_Regressed_netResidualExpression.tsv', name = 'Normalised, covariates and Diagnosis removed residual expression (for network analysis)', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
  all.annotations$dataSubType = 'residualGeneExpForNetAnlz'
  syn_temp$setAnnotations(ENRICH_OBJ, annotations = all.annotations)
```

### R Source Code
[Github](`r thisFile`)

```{r knitmd, eval=FALSE, cache=FALSE, include=FALSE}
reticulate::use_python("/usr/bin/python", required = TRUE)
synapseclient <- reticulate::import("synapseclient")
syn_temp <- synapseclient$Synapse()
syn_temp$login()

setwd("~/TWAS/code/")
source("~/TWAS/utilityFunctions/knitfile2synapseClient.R")
source("~/TWAS/utilityFunctions/hook_synapseMdSyntax_plot.R")
createAndKnitToFolderEntityClient(file = "",
                                          parentId ="syn21136908",
                                          folderName = 'CMC_AnalysisExp')
```